src/AspNet.Security.OAuth.Alipay/AlipayAuthenticationHandler.cs

```diff
@@ -0,0 +1,231 @@
+                            /* Message: */ payload.RootElement.GetString("msg"));
+
+                throw new HttpRequestException("An error occurred while retrieving user information.");
+            }
+
+            identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, payload.RootElement.GetString("user_id"), ClaimValueTypes.String, Options.ClaimsIssuer));
+
+            var principal = new ClaimsPrincipal(identity);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+
+            context.RunClaimActions();
+
+            await Options.Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
+        /// <summary>
+        /// Gets an HTTP payload as an asynchronous operation.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        /// <param name="elementName">Name of the element.</param>
+        /// <returns>Task&lt;JsonDocument&gt;.</returns>
+        private async Task<JsonDocument> ParseJsonAsync(HttpResponseMessage message, string elementName)
+        {
+            var messageContent = await message.Content.ReadAsStringAsync();
+
+            var document = JsonDocument.Parse(messageContent);
+            var payload = JsonDocument.Parse(document.RootElement.GetString(elementName));
+
+            return payload;
+        }
+
+        /// <summary>
+        /// Gets the RSA2 signature.
+        /// </summary>
+        /// <param name="source">The source.</param>
+        /// <returns>System.String.</returns>
+        private string GetRSA2Signature(SortedDictionary<string, string> source)
+        {
+            var builder = new StringBuilder();
+
+            foreach (var pair in source)
+            {
+                if (string.IsNullOrEmpty(pair.Value))
+                {
+                    continue;
+                }
+
+                builder.Append(pair.Key)
+                       .Append('=')
+                       .Append(pair.Value)
+                       .Append('&');
+            }
+
+            builder.Remove(builder.Length - 1, 1);
+
+            byte[] dataBytes = Encoding.UTF8.GetBytes(builder.ToString());
+
+            using var privateKeyRsaProvider = CreateAlgorithm();
+            byte[] signatureBytes = privateKeyRsaProvider.SignData(dataBytes, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
+
+            return Convert.ToBase64String(signatureBytes);
+        }
+
+        private RSA CreateAlgorithm()
+        {
+            byte[] privateKeyBytes = Convert.FromBase64String(Options.ClientSecret);
+            var algorithm = RSA.Create();
+
+            try
+            {
+                algorithm.ImportRSAPrivateKey(privateKeyBytes, out int _);
+                return algorithm;
+            }
+            catch (Exception)
+            {
+                algorithm?.Dispose();
```

Comment:
I'd remove the `?` as there's no reason for `algorithm` to be null here.

Reply:
Ah, my bad. This is copy-pasta from the Apple provider that must’ve got leftover after a refactor.

Reply:
Ah, my bad. This is copy-pasta from the Apple provider that must’ve got leftover after a refactor.